File: ios/Sources/SilverCareCore/FallDetectorCore.swift
import Foundation

public struct FallDetectorDefaults: Equatable {
    public var impactAccelerationThreshold: Double
    public var impactDeviationThreshold: Double
    public var rotationThreshold: Double
    public var visualDiffStrong: Double
    public var visualBrightnessRangeStrong: Double
    public var visualSpikeCountStrong: Int
    public var minimumVisualSamples: Int

    public init(
        impactAccelerationThreshold: Double = 27,
        impactDeviationThreshold: Double = 17,
        rotationThreshold: Double = 260,
        visualDiffStrong: Double = 0.20,
        visualBrightnessRangeStrong: Double = 0.40,
        visualSpikeCountStrong: Int = 3,
        minimumVisualSamples: Int = 6
    ) {
        self.impactAccelerationThreshold = impactAccelerationThreshold
        self.impactDeviationThreshold = impactDeviationThreshold
        self.rotationThreshold = rotationThreshold
        self.visualDiffStrong = visualDiffStrong
        self.visualBrightnessRangeStrong = visualBrightnessRangeStrong
        self.visualSpikeCountStrong = visualSpikeCountStrong
        self.minimumVisualSamples = minimumVisualSamples
    }
}

public struct MotionSensorSample: Equatable {
    public var time: TimeInterval
    public var gravity: SIMD3<Double>
    public var accelerationMagnitude: Double
    public var accelerationDeviation: Double
    public var rotationMagnitude: Double

    public init(
        time: TimeInterval = 0,
        gravity: SIMD3<Double>,
        accelerationMagnitude: Double,
        accelerationDeviation: Double,
        rotationMagnitude: Double
    ) {
        self.time = time
        self.gravity = gravity
        self.accelerationMagnitude = accelerationMagnitude
        self.accelerationDeviation = accelerationDeviation
        self.rotationMagnitude = rotationMagnitude
    }
}

public struct VisualFrameSample: Equatable {
    public var time: TimeInterval
    public var diff: Double
    public var brightness: Double

    public init(time: TimeInterval, diff: Double, brightness: Double) {
        self.time = time
        self.diff = diff
        self.brightness = brightness
    }
}

public struct VisualEvidence: Equatable {
    public var sampleCount: Int
    public var maxDiff: Double
    public var spikeCount: Int
    public var brightnessRange: Double

    public init(sampleCount: Int, maxDiff: Double, spikeCount: Int, brightnessRange: Double) {
        self.sampleCount = sampleCount
        self.maxDiff = maxDiff
        self.spikeCount = spikeCount
        self.brightnessRange = brightnessRange
    }
}

public enum FallDetectorCore {
    public static func motionSample(
        gravity: SIMD3<Double>,
        rotation: SIMD3<Double>,
        time: TimeInterval = 0
    ) -> MotionSensorSample {
        let accelerationMagnitude = vectorLength(gravity)
        let accelerationDeviation = abs(accelerationMagnitude - 9.81)
        let rotationMagnitude = vectorLength(rotation)
        return MotionSensorSample(
            time: time,
            gravity: gravity,
            accelerationMagnitude: accelerationMagnitude,
            accelerationDeviation: accelerationDeviation,
            rotationMagnitude: rotationMagnitude
        )
    }

    public static func hasFallImpact(
        _ sample: MotionSensorSample,
        defaults: FallDetectorDefaults = FallDetectorDefaults()
    ) -> Bool {
        sample.accelerationMagnitude >= defaults.impactAccelerationThreshold
            || sample.accelerationDeviation >= defaults.impactDeviationThreshold
            || sample.rotationMagnitude >= defaults.rotationThreshold
    }

    public static func nextBaselineGravity(
        current: SIMD3<Double>?,
        sensor: MotionSensorSample,
        impactActive: Bool
    ) -> SIMD3<Double>? {
        guard !impactActive else { return current }
        guard sensor.accelerationDeviation < 1.2, sensor.rotationMagnitude < 12 else {
            return current
        }
        guard let current else { return sensor.gravity }
        return current * 0.88 + sensor.gravity * 0.12
    }

    public static func visualEvidence(
        samples: [VisualFrameSample],
        now: TimeInterval,
        window: TimeInterval
    ) -> VisualEvidence {
        let recent = samples.filter { now - $0.time <= window }
        guard !recent.isEmpty else {
            return VisualEvidence(sampleCount: 0, maxDiff: 0, spikeCount: 0, brightnessRange: 0)
        }
        let diffs = recent.map(\.diff)
        let brightness = recent.map(\.brightness)
        return VisualEvidence(
            sampleCount: recent.count,
            maxDiff: diffs.max() ?? 0,
            spikeCount: diffs.filter { $0 >= 0.12 }.count,
            brightnessRange: (brightness.max() ?? 0) - (brightness.min() ?? 0)
        )
    }

    public static func isVisualEvidenceStrong(
        _ evidence: VisualEvidence,
        defaults: FallDetectorDefaults = FallDetectorDefaults()
    ) -> Bool {
        evidence.sampleCount >= defaults.minimumVisualSamples
            && evidence.maxDiff >= defaults.visualDiffStrong
            && (
                evidence.spikeCount >= defaults.visualSpikeCountStrong
                || evidence.brightnessRange >= defaults.visualBrightnessRangeStrong
            )
    }

    public static func shouldConfirmFall(
        sensor: (maxAcceleration: Double, maxDeviation: Double, maxRotation: Double),
        visual: VisualEvidence,
        defaults: FallDetectorDefaults = FallDetectorDefaults()
    ) -> Bool {
        let sensorStrong = sensor.maxAcceleration >= defaults.impactAccelerationThreshold
            || sensor.maxDeviation >= defaults.impactDeviationThreshold
            || sensor.maxRotation >= defaults.rotationThreshold
        return sensorStrong && isVisualEvidenceStrong(visual, defaults: defaults)
    }

    public static func isRecoveredFromFall(
        baselineGravity: SIMD3<Double>,
        currentGravity: SIMD3<Double>,
        sensor: MotionSensorSample
    ) -> Bool {
        angleBetween(baselineGravity, currentGravity) < 12
            && sensor.accelerationDeviation < 2.2
            && sensor.rotationMagnitude < 20
    }

    public static func angleBetween(_ lhs: SIMD3<Double>, _ rhs: SIMD3<Double>) -> Double {
        let denominator = max(0.0001, vectorLength(lhs) * vectorLength(rhs))
        let dotProduct = lhs.x * rhs.x + lhs.y * rhs.y + lhs.z * rhs.z
        let cosine = min(1, max(-1, dotProduct / denominator))
        return acos(cosine) * 180 / .pi
    }

    private static func vectorLength(_ value: SIMD3<Double>) -> Double {
        sqrt(value.x * value.x + value.y * value.y + value.z * value.z)
    }
}

File: app/src/main/assets/static/js/fall_detector_core.js
export const FALL_DEFAULTS = {
    sampleIntervalMs: 250,
    bufferWindowMs: 4500,
    sampleWidth: 72,
    sampleHeight: 54,
    probeDelayMs: 900,
    cooldownMs: 24000,
    countdownSeconds: 10,
    impactDeviation: 15,
    impactMagnitude: 24,
    rotationImpact: 220,
    visualDiffSpike: 0.13,
    visualDiffStrong: 0.18,
    brightnessRangeStrong: 0.38,
    recoveryAngle: 25,
    recoveryDeviation: 4,
    recoveryRotation: 35,
    recoveryHoldMs: 1800
};

export function clampFallSensitivity(value) {
    const level = Number.parseInt(value, 10);
    if (!Number.isFinite(level)) return 10;
    return Math.max(1, Math.min(10, level));
}

export function fallConfigForSensitivity(value, base = FALL_DEFAULTS) {
    const level = clampFallSensitivity(value);
    const sensitivityRatio = (level - 1) / 9;
    const thresholdMultiplier = 1 + (1 - sensitivityRatio) * 1.4;
    return {
        ...base,
        sensitivityLevel: level,
        impactDeviation: Number((base.impactDeviation * thresholdMultiplier).toFixed(2)),
        impactMagnitude: Number((base.impactMagnitude * thresholdMultiplier).toFixed(2)),
        rotationImpact: Number((base.rotationImpact * thresholdMultiplier).toFixed(2)),
        visualDiffSpike: Number(Math.min(0.42, base.visualDiffSpike * thresholdMultiplier).toFixed(3)),
        visualDiffStrong: Number(Math.min(0.5, base.visualDiffStrong * thresholdMultiplier).toFixed(3)),
        brightnessRangeStrong: Number(Math.min(0.8, base.brightnessRangeStrong * thresholdMultiplier).toFixed(3))
    };
}

export function readSensorFromMotion(event, time = Date.now()) {
    const gravity = event.accelerationIncludingGravity || {};
    const x = Number(gravity.x);
    const y = Number(gravity.y);
    const z = Number(gravity.z);
    if (!Number.isFinite(x) || !Number.isFinite(y) || !Number.isFinite(z)) return null;

    const rotation = event.rotationRate || {};
    const alpha = Number(rotation.alpha) || 0;
    const beta = Number(rotation.beta) || 0;
    const gamma = Number(rotation.gamma) || 0;
    const accMagnitude = Math.sqrt(x * x + y * y + z * z);
    const rotationMagnitude = Math.sqrt(alpha * alpha + beta * beta + gamma * gamma);

    return {
        time,
        gravity: { x, y, z },
        accMagnitude,
        accDeviation: Math.abs(accMagnitude - 9.81),
        rotationMagnitude
    };
}

export function hasFallImpact(sensor, config = FALL_DEFAULTS) {
    return sensor.accDeviation >= config.impactDeviation
        || sensor.accMagnitude >= config.impactMagnitude
        || sensor.rotationMagnitude >= config.rotationImpact;
}

export function nextBaselineGravity(previous, sensor, alertActive = false) {
    if (sensor.accDeviation > 2.8 || sensor.rotationMagnitude > 45 || alertActive) return previous;
    if (!previous) return { ...sensor.gravity };

    const keep = 0.92;
    const add = 1 - keep;
    return {
        x: previous.x * keep + sensor.gravity.x * add,
        y: previous.y * keep + sensor.gravity.y * add,
        z: previous.z * keep + sensor.gravity.z * add
    };
}

export function computeVisualEvidence(samples, windowMs, now = Date.now(), config = FALL_DEFAULTS) {
    const cutoff = now - windowMs;
    const activeSamples = samples.filter((sample) => sample.time >= cutoff);
    let maxDiff = 0;
    let spikeCount = 0;
    let minBrightness = 1;
    let maxBrightness = 0;

    activeSamples.forEach((sample) => {
        maxDiff = Math.max(maxDiff, sample.diff);
        if (sample.diff >= config.visualDiffSpike) spikeCount += 1;
        minBrightness = Math.min(minBrightness, sample.brightness);
        maxBrightness = Math.max(maxBrightness, sample.brightness);
    });

    return {
        sampleCount: activeSamples.length,
        maxDiff: Number(maxDiff.toFixed(3)),
        spikeCount,
        brightnessRange: Number(Math.max(0, maxBrightness - minBrightness).toFixed(3))
    };
}

export function isVisualEvidenceStrong(visual, config = FALL_DEFAULTS) {
    return visual.sampleCount >= 5
        && (
            visual.maxDiff >= config.visualDiffStrong
            || visual.spikeCount >= 2
            || visual.brightnessRange >= config.brightnessRangeStrong
        );
}

export function shouldConfirmFall(sensorEvidence, visualEvidence, config = FALL_DEFAULTS) {
    const sensorStrong = !!sensorEvidence && (
        sensorEvidence.maxDeviation >= config.impactDeviation
        || sensorEvidence.maxAcc >= config.impactMagnitude
        || sensorEvidence.maxRotation >= config.rotationImpact
    );
    return sensorStrong && isVisualEvidenceStrong(visualEvidence, config);
}

export function angleBetween(a, b) {
    const dot = a.x * b.x + a.y * b.y + a.z * b.z;
    const am = Math.sqrt(a.x * a.x + a.y * a.y + a.z * a.z);
    const bm = Math.sqrt(b.x * b.x + b.y * b.y + b.z * b.z);
    if (!am || !bm) return 180;
    const cos = Math.max(-1, Math.min(1, dot / (am * bm)));
    return Math.acos(cos) * 180 / Math.PI;
}

export function isRecoveredFromFall(baseline, currentGravity, sensor, config = FALL_DEFAULTS) {
    if (!baseline || !currentGravity || !sensor) return false;
    const angle = angleBetween(baseline, currentGravity);
    return angle <= config.recoveryAngle
        && sensor.accDeviation <= config.recoveryDeviation
        && sensor.rotationMagnitude <= config.recoveryRotation;
}
